ragatouille/models/torch_kmeans.py

```diff
@@ -0,0 +1,119 @@
+import time
+
+import torch
+
+
+def _train_kmeans(self, sample, shared_lists):  # noqa: ARG001
+    if self.use_gpu:
+        torch.cuda.empty_cache()
+    centroids = compute_pytorch_kmeans(
+        sample,
+        self.config.dim,
+        self.num_partitions,
+        self.config.kmeans_niters,
+        self.use_gpu,
+    )
+    centroids = torch.nn.functional.normalize(centroids, dim=-1)
+    if self.use_gpu:
+        centroids = centroids.half()
+    else:
+        centroids = centroids.float()
+    return centroids
+
+
+def compute_pytorch_kmeans(
+    sample,
+    dim,
+    num_partitions,
+    kmeans_niters,
+    use_gpu,
+    batch_size=512000,
+    tol=1e-4,
+    verbose=1,
+):
+    device = torch.device("cuda" if use_gpu else "cpu")
+    sample = sample.to(device)
+    total_size = sample.shape[0]
+
+    # Initialize centroids randomly
+    centroids = torch.randn(num_partitions, dim, dtype=sample.dtype, device=device)
+
+    # Convert to half-precision if GPU is available
+    if use_gpu:
+        sample = sample.half()
+        centroids = centroids.half()
+    else:
+        sample = sample.float()
+        centroids = centroids.float()
+
+    # Precompute the squared norms of data points
+    sample_norms = torch.sum(sample.pow(2), dim=1, keepdim=True)
+
+    start_time = time.time()
+    for i in range(kmeans_niters):
+        iter_time = time.time()
+
+        # Shuffle the data points
+        permutation = torch.randperm(total_size, device=device)
+        sample = sample[permutation]
+        sample_norms = sample_norms[permutation]
+
+        if total_size <= batch_size:
```

Comment:
Nitpick: Could we perhaps merge both branches here? AFAICT they only differ in the way the centroids are computed. Maybe just introduce `lag = 0 if total_size <= batch_size else 0.9` before L97.

Reply:
Out of curiosity: is there any rationale behind the concrete value of 0.9?

Reply:
> Nitpick: Could we perhaps merge both branches here? AFAICT they only differ in the way the centroids are computed. Maybe just introduce lag = 0 if total_size <= batch_size else 0.9 before L97.

Yes that'd make complete sense.

> Out of curiosity: is there any rationale behind the concrete value of 0.9?

No scientific one at all, to be honest! I've often found that it empirically works well to have smooth updates but this empirical evidence is all I have 😄